Avro message for phone-home ping response.

apovzner opened the pull request:
Intend to use it at least for KSQL Beacon and C3 phone-home.

KSQL Beacon currently assumes plaintext response with OK status. The plan is to use "Accept" header, where no "Accept" header would mean "text/plain" for KSQL pings, and otherwise `PhoneHomeResponse` json response for successful pings (ErrorMessage pings for errors).

apovzner commented:
@ewencp The only reason I chose Avro is because proactive support/`support-metrics-common` already uses Avro for requests to the server. I was thinking about using accept header to tell the server which response type the client expects: right now will be plaintext vs. avro for KSQL Beacon. (vs. "api-version" we will use for versioning requests to the server )

What do you think is a better option? I think it should be ok to use something different for responses.

ewencp commented:
It's just not a very common format to use for REST APIs, but I didn't really have context from the existing implementation. For better or worse, most people would expect JSON.

I don't feel super strongly about it though. I'm not sure how the output is actually going to be used, but that probably affects the choice significantly. If it is expected to be human readable, Avro would be a pain. If we're only consuming from applications, then it is less of an issue.

apovzner commented:
@ewencp Could you please look at this PR again? I agree that avro somehow looks strange in a response. So, I replaced avro message with json that KSQL Beacon can use and I think C3, and also json for error response. I also added a new `ConfluentResponseHandler` interface that returns media type for "Accept" header. This way, new clients can define their own response messages, including avro if they want. They can also add a version to the "accept" header.

The path to move KSQL Beacon to use JSON instead of `text/plain` response type (which I already tested locally): the server will check the "accept" header for the ksql resource. If there is no "accept" header in ksql message, then it means that the server should return "text/plain" as a response (only OK response, KSQL does not handle error responses right now, so we can safely move to JSON error responses). Otherwise, the server will return `PhoneHomeResponse`.

@norwood please look if `PhoneHomeResponse` for success and `ErrorMessage` for error would work for C3 phone-home. If we don't know what response we need from the server in case of success, we can either for now use `PhoneHomeResponse` (the server will for now set all the fields to null, until we decide what it should send for the client to log) or not handle any successful response for now. If we use `PhoneHomeResponse`, I propose to pass `ConfluentResponseHandler` to `ConfluentSubmitter` (instead of `ResponseHandler`) and return something like "application/vnd.phonehome.v1+json" from `getAcceptMediaType()`, so that we can define another json if we want something completely different from the response.

ewencp commented on support-metrics-common/src/main/java/io/confluent/support/metrics/response/ErrorResponse.java:
```diff
@@ -0,0 +1,53 @@
+
+  @JsonCreator
+  public ErrorResponse(
+      @JsonProperty("message") String message) {
```
nit: formatting looks off here

ewencp commented on support-metrics-common/src/main/java/io/confluent/support/metrics/response/PhoneHomeResponse.java:
```diff
@@ -0,0 +1,66 @@
+  @JsonCreator
+  public PhoneHomeResponse(
+      @JsonProperty("confluentPlatformVersion") String confluentPlatformVersion,
+      @JsonProperty("informationForUser") String informationForUser) {
```
nit: we should figure out standardization of this / checkstyle, but elsewhere when we break arg lists across lines we put the final `) {` on its own line

ewencp commented on support-metrics-common/src/main/java/io/confluent/support/metrics/submitters/ConfluentResponseHandler.java:
```diff
@@ -0,0 +1,23 @@
+
+package io.confluent.support.metrics.submitters;
+
+public interface ConfluentResponseHandler extends ResponseHandler {
```
Does this actually need a separate interface or can we extend `ResponseHandler`?

apovzner replied:
After a few days, a separate interface does not look like a good solution anymore. The reason I added it (vs. adding new method to `ResponseHandler` and making all existing clients implement it) was because i did not want to break builds across repos, especially as we are in the middle of the release. So I am going to remove this interface from this PR, and then change ResponseHandler after the CP release. We don't need it right now for C3, because we can  assume JSON responses for C3. We can also replace all error responses in the server to JSON (common code which is also used by KSQL), because none of existing clients (=KSQL and proactive support) check error response right now.  And leave KSQL success response as text for a bit longer.
